modules/renter/proto/negotiate.go

```diff
@@ -114,3 +114,116 @@ func verifyRecentRevision(conn net.Conn, contract Contract) error {
 	// contract expiration (which was checked earlier).
 	return modules.VerifyFileContractRevisionTransactionSignatures(lastRevision, hostSignatures, contract.FileContract.WindowStart-1)
 }
+
+// negotiateRevision sends a revision and actions to the host for approval,
+// completing one iteration of the revision loop.
+func negotiateRevision(conn net.Conn, rev types.FileContractRevision, secretKey crypto.SecretKey) (types.Transaction, error) {
+	// create transaction containing the revision
+	signedTxn := types.Transaction{
+		FileContractRevisions: []types.FileContractRevision{rev},
+		TransactionSignatures: []types.TransactionSignature{{
+			ParentID:       crypto.Hash(rev.ParentID),
+			CoveredFields:  types.CoveredFields{FileContractRevisions: []uint64{0}},
+			PublicKeyIndex: 0, // renter key is always first -- see formContract
+		}},
+	}
+	// sign the transaction
+	encodedSig, _ := crypto.SignHash(signedTxn.SigHash(0), secretKey) // no error possible
+	signedTxn.TransactionSignatures[0].Signature = encodedSig[:]
+
+	// send the revision
+	if err := encoding.WriteObject(conn, rev); err != nil {
+		return types.Transaction{}, errors.New("couldn't send revision: " + err.Error())
+	}
+	// read acceptance
+	if err := modules.ReadNegotiationAcceptance(conn); err != nil {
+		return types.Transaction{}, errors.New("host did not accept revision: " + err.Error())
+	}
+
+	// send the new transaction signature
+	if err := encoding.WriteObject(conn, signedTxn.TransactionSignatures[0]); err != nil {
+		return types.Transaction{}, errors.New("couldn't send transaction signature: " + err.Error())
+	}
+	// read the host's acceptance and transaction signature
+	if err := modules.ReadNegotiationAcceptance(conn); err != nil {
+		return types.Transaction{}, errors.New("host did not accept transaction signature: " + err.Error())
+	}
+	var hostSig types.TransactionSignature
+	if err := encoding.ReadObject(conn, &hostSig, 16e3); err != nil {
+		return types.Transaction{}, errors.New("couldn't read host's signature: " + err.Error())
+	}
+
+	// add the signature to the transaction and verify it
+	// NOTE: we can fake the blockheight here because it doesn't affect
+	// verification; it just needs to be above the fork height and below the
+	// contract expiration (which was checked earlier).
```

Comment:
`StandaloneValid` is influenced by an exact height, because we changed the fee at I think block 21,000. In production, it's not significant but if you aren't providing an exact height you might see some weird things in testing if the height is close to testing fork height.